### PR TITLE
storage: bounded memory upsert snapshot

### PR DESCRIPTION
**Pull request by guswynn**
This replaces: https://github.com/MaterializeInc/materialize/pull/18912

This pr is a complex one: at a high level, other than tests, there are 4 things going on:

- A new method on `UpsertState` that asks implementation to consolidate _batches of updates (including diffs)_ in place. In the case of the rocksdb implementation, this does in fact mean _bounded-memory_
  - There is significant machinery around plugging this api in.
- A large amount of machinery to maintain the the statistics added in https://github.com/MaterializeInc/materialize/pull/19044. 
  - This requires some copied code between the implementations, as they store values encoded differently.
- The actual consolidation logic. Instead of storing a `Key->List[Update]`, we instead leverage the power of `xor`, as described below. Note that this logic is primarily encapsulated in the `StateValue` struct, and is used by both implementations.
- The fairly complex rocksdb implementation of the new API. I intended this to be a simple as possible, but it is still a significant lift.

As for the xor logic itself, this description is copied directly from @petrosagg in the original base of this pr (https://github.com/MaterializeInc/materialize/pull/18912):

This makes recovering the upsert snapshot use memory proportinal to the
number of keys as opposed to the current method that uses memory
proportional to the total number of updates.

We use a XOR trick in order to accumulate the snapshot without having to
store the full unconsolidated history in memory. For all (value, diff)
updates of a key we track:
- diff_sum = SUM(diff)
- value_accum = XOR(bincode(value))
- checksum_accum = XOR(checksum(bincode(value)))

The method is correct because a well formed upsert snapshot will have
for each key:
- Zero or one updates of the form (cur_value, +1)
- Zero or more pairs of updates of the form (prev_value, +1), (prev_value, -1)

We are interested in extracting the cur_value of each key and discard
all prev_values that might be included in the stream. Since the history
of prev_values always comes in pairs, computing the XOR of those is
always going to cancel their effects out. Also, since XOR is commutative
this property is true independent of the order.

Therefore the accumulators will end up precisely in one of two states:
1. diff == 0, checksum == 0, value == [0..] => the key is not present
2. diff == 1, checksum == checksum(cur_value) value == cur_value => the
   key is present

In the absense of bugs, accumulating the diff and checksum is not
required since we know that a well formed snapshot always satisfies
XOR(bincode(values)) == bincode(cur_value). However bugs may happen and
so storing 16 more bytes per key to have a very high guarantee that
we're not decoding garbage is more than worth it.

### Motivation

  * This PR adds a known-desirable feature.
 
 Part of: https://github.com/MaterializeInc/materialize/issues/17067

### Tips for reviewer

The bottom commit all the logic. The top commit contains additional tests.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
in notion
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

### CODE REVIEW

**Review by philip-stoev — APPROVED** [review]
Thank you for running Nightly. The "limits" failure is unrelated to this PR.

As I mentioned in @petrosagg's original PR, I am not sure what else is there to test about this PR. Please let me know if you have any ideas.

File: src/storage/src/render/upsert/types.rs [hunk 1]
```diff
@@ -7,85 +7,478 @@
...
+    /// We are interested in extracting the cur_value of each key and discard all prev_values
+    /// that might be included in the stream. Since the history of prev_values always comes in
+    /// pairs, computing the XOR of those is always going to cancel their effects out. Also,
+    /// since XOR is commutative this property is true independent of the order. The same is
+    /// true for the summations of the length and checksum since the sum will contain the
+    /// unrelated values zero times.
+    ///
+    /// Therefore the accumulators will end up precisely in one of two states:
+    /// 1. diff == 0, checksum == 0, value == [0..] => the key is not present
+    /// 2. diff == 1, checksum == checksum(cur_value) value == cur_value => the key is present
```
Comment (moulimukherjee) [inline src/storage/src/render/upsert/types.rs:155]: nit: Can a comment be added for the expected return value? That it will return `false` if a key was effectively added and `true` otherwise?

File: src/storage/src/render/upsert/types.rs [hunk 2]
```diff
@@ -7,85 +7,478 @@
...
+                .unwrap();
+            let len = i64::try_from(bincode_buffer.len()).unwrap();
+
+            *diff_sum += diff;
+            *len_sum += len.wrapping_mul(diff);
+            // Truncation is fine (using `as`) as this is just a checksum
+            *checksum_sum += (seahash::hash(&*bincode_buffer) as i64).wrapping_mul(diff);
+
+            // XOR of even diffs cancel out, so we only do it if diff is odd
+            if diff.abs() % 2 == 1 {
```
Comment (moulimukherjee) [inline src/storage/src/render/upsert/types.rs:191]: dumb q: Will `diff` for the single value ever be **not** `+1` or `-1`?

Comment (moulimukherjee) [reply src/storage/src/render/upsert/types.rs:191]: Actually should this be comparing `diff_sum` instead of `diff`?

Comment (guswynn) [reply src/storage/src/render/upsert/types.rs:191]: Not in the source pipeline!

File: src/storage/src/render/upsert/types.rs [hunk 3]
```diff
@@ -7,85 +7,478 @@
...
+            stats.processed_puts += 1;
+            match p_value.value {
                 Some(value) => {
-                    self.state.insert(key, value);
+                    let decoded_value = StateValue::Decoded(value);
+                    let size: i64 = decoded_value
+                        .memory_size()
+                        .try_into()
+                        .expect("less than i64 size");
+                    match p_value.previous_persisted_size {
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/render/upsert/types.rs:441]: No, because the `envelope_state_bytes` will be seeded with the original (larger) size. `multi_put` is allowed to return negative numbers of the size has gone down (cause updates are smaller, or lots of deletes)

**Comment by guswynn** [comment]
@philip-stoev The best thing to test would be rehydrating a large upsert source after killing a clusterd, which i intend to add in a followup ci pr

File: src/storage/src/render/upsert/types.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (moulimukherjee) [reply src/storage/src/render/upsert/types.rs:441]: Say, we have only one value in the state and say it's size when `Snapshotting` is 1 and when `Decoded`, it's 2.

Originally `envelope_state_bytes` will have 1. If the record is now removed, we'll be subtracting 2 since it's in `Decoded` form. Overall `envelope_state_bytes` is now weirdly -1.

File: src/storage/src/render/upsert.rs [hunk 4]
```diff
@@ -312,18 +246,37 @@ where
...
+                        #[allow(clippy::as_conversions)]
+                        if !resume_upper.less_equal(&ts) {
+                            update_buf.push((key, value, diff));
+                            batch_key_counter.insert(key);
+
+                            // If the number of keys in this batch is >= the requested
+                            // size, flush it out. We also flush if there are >= `SNAPSHOT_BATCH_SIZE`*10
+                            // updates, with < SNAPSHOT_BATCH_SIZE keys.
+                            if batch_key_counter.len() >= US::SNAPSHOT_BATCH_SIZE
+                                || update_buf.len() >= US::SNAPSHOT_BATCH_SIZE * 10
```
Comment (petrosagg) [inline src/storage/src/render/upsert.rs:270]: I'm not sure if this is worth doing. When an operator runs in timely there is a certain amount of data ready to go in a handle, stored in memory. What we want to do is drain all the data in the input handle and process it in the upsert store, which you can do by calling `now_or_never()` on the input handle. With the fixed batch size you're risking accumulating work in memory that could be retired without touching the hashmap at all.

Ideally the `AsyncInputHandle` should have a `.ready()` and `next_sync()` API to enable this pattern without resorting to `now_or_never()` but I suppose you don't want to do this in this PR (but if you do I fully support it!)

File: src/storage/src/render/upsert.rs [hunk 5]
```diff
@@ -332,54 +285,54 @@ where
...
             .rehydration_total
-            .set(u64::cast_from(snapshot_size));
-
+            .set(
+                stats
+                    .values_diff
+                    .try_into()
+                    .unwrap_or_else(|e: std::num::TryFromIntError| {
+                        tracing::warn!(
+                            "rehydration_total metric overflowed and is innacurate: {}",
```
Comment (petrosagg) [inline src/storage/src/render/upsert.rs:308]: Here we convert from `i64` to `u64` which only fails when `values_diff` is negative and not due to overflow. When does this happen? Should we just return 0 or panic because there is a bug?

File: src/storage/src/render/upsert.rs [hunk 6]
```diff
@@ -332,54 +285,54 @@ where
...
+                            e.display_with_causes(),
+                        );
+
+                        0
+                    }),
+            );
+
+        // These `set_` functions also ensure that these values are non-negative.
```
Comment (petrosagg) [inline src/storage/src/render/upsert.rs:316]: Same question here. When do negative values happen?

File: src/storage/src/render/upsert.rs [hunk 7]
```diff
@@ -332,54 +285,54 @@ where
...
-            .set_envelope_state_count(initial_state_records);
+            .set_envelope_state_count(stats.values_diff);
+
+        drop(previous_token);
+        while let Some(_event) = previous.next().await {
+            // Exchaust the previous input. It is expected to immediately reach the empty
+            // antichain since we have dropped its token.
+        }
+
+        // A re-usable buffer of changes, per key. This is
```
Comment (petrosagg) [inline src/storage/src/render/upsert.rs:330]: Can you justify to 100 (or 80 if you're that kind of person) characters?

File: src/storage/src/render/upsert.rs [hunk 8]
```diff
@@ -435,41 +380,67 @@ where
...
                     // value (or deletion). These will be synced at the end to the `state`.
+                    //
+                    // Note that we are effectively doing "mini-upsert" here, using
+                    // `command_state`. This "mini-upsert" is seeded with data from `state`, using
+                    // a single `multi_get` above, and the final state is written out into
+                    // `state` using a single `multi_put`. This simplifies `UpsertState`
+                    // implementations, and reduces the number of reads and write we need to do.
+                    //
+                    // This "mini-upsert" technique is actually useful in some `UpsertState`
+                    // implementations, themselves, see `types::rocksdb` for an example. In some
```
Comment (petrosagg) [inline src/storage/src/render/upsert.rs:394]: What does this comment mean? The `UpsertState` implementation should never do this logic and just blindly write what we tell them from here, right?

File: src/storage/src/render/upsert/rocksdb.rs [hunk 9]
```diff
@@ -24,51 +27,186 @@ pub const BATCH_SIZE: usize = 1024;
 /// A `UpsertState` implementation backed by RocksDB.
 /// This is currently untested, and simply compiles.
 pub struct RocksDB {
-    rocksdb: RocksDBInstance<UpsertKey, UpsertValue>,
+    inner: Inner,
+
+    // Bincode options and buffer used
+    // in `merge_snapshot_chunk`.
+    bincode_opts: BincodeOpts,
```
Comment (petrosagg) [inline src/storage/src/render/upsert/rocksdb.rs:34]: Can you explain why the XOR details have to leak into the implementation of key/val stores? I was expecting that all this logic would be done externally and these implementation would only implement a dumb `multi_get`/`multi_put` API

File: src/storage/src/render/upsert/types.rs [hunk 10]
```diff
@@ -7,85 +7,478 @@
...
+/// The default set of `bincode` options used for consolidating
+/// upsert snapshots (and writing values to RocksDB).
+pub type BincodeOpts = bincode::config::WithOtherTrailing<
+    bincode::config::DefaultOptions,
+    bincode::config::AllowTrailing,
+>;
+
+/// Build the default `BincodeOpts`.
+pub fn upsert_bincode_opts() -> BincodeOpts {
+    bincode::DefaultOptions::new().allow_trailing_bytes()
```
Comment (petrosagg) [inline src/storage/src/render/upsert/types.rs:87]: We no longer need this because we track the length of the XOR buffer so we can just the normal bincode options. In my original prototype I didn't track the length which is why I had to do this

File: src/storage/src/render/upsert/types.rs [hunk 11]
```diff
@@ -7,85 +7,478 @@
...
+}
+
+pub type UpsertValueAndSize = mz_rocksdb::GetResult<StateValue>;
+
+#[derive(Clone, Debug)]
+pub struct PutValue {
```
Comment (petrosagg) [inline src/storage/src/render/upsert/types.rs:93]: Similarly to `GetValue` I think the options should move outside and use `Option<PutValue>` everywhere that is needed so that the type only encodes valid values

File: src/storage/src/render/upsert/types.rs [hunk 12]
```diff
@@ -7,85 +7,478 @@
...
+    Decoded(UpsertValue),
+}
+
+/// A value as produced during consolidation of a snapshot.
+#[derive(Clone, Default, serde::Serialize, serde::Deserialize, Debug)]
+pub struct Snapshotting {
+    value_xor: Vec<u8>,
+    len_sum: Wrapping<i64>,
+    checksum_sum: Wrapping<i64>,
+    diff_sum: Wrapping<i64>,
```
Comment (petrosagg) [inline src/storage/src/render/upsert/types.rs:123]: `std::num::Wrapping`, very nice!

File: src/storage/src/render/upsert/types.rs (hunk 1 above)
Comment (petrosagg) [reply src/storage/src/render/upsert/types.rs:155]: Ditto, I wondered the same. The return value seems to indicate if the value if effectively deleted after the merge operation

File: src/storage/src/render/upsert/types.rs [hunk 13]
```diff
@@ -7,85 +7,478 @@
...
+            }
+
+            return diff_sum.0 == 0 && checksum_sum.0 == 0 && value_xor.iter().all(|&x| x == 0);
+        } else {
+            panic!("`merge_update` called after snapshot consolidation")
+        }
+    }
+
+    /// After consolidation of a snapshot, we assume that all values in the `UpsertState` implementation
+    /// are `Self::Snapshotting`, with a `diff_sum` of 1. Afterwards, if we need to retract one of
```
Comment (petrosagg) [inline src/storage/src/render/upsert/types.rs:207]: This comment is slightly inaccurate since the implementation below also allows `diff_sum` to be `0` provided that everything else is zero.

**Review by petrosagg — COMMENTED** [review]
This looks good overall but I have a few high level concerns that we should discuss:

1. The XOR logic is leaking into each individual key/value implementation. That seems problematic to me as swapping the key/value store is no longer trivial or at least only concerned with the details of the store. Why does it have to be that way?
2. Metric related things seem to have infiltrated every single type, carrying around decoded bytes, counts, and interleaving with the actual logic. This makes the code hard to follow, is there an alternative solution where we pass a `Metrics` registry or something to the implementations where they can set their metrics directly?

File: src/storage/src/render/upsert.rs (hunk 6 above)
Comment (moulimukherjee) [reply src/storage/src/render/upsert.rs:316]: When rehydrating state, the `envelope_state_bytes` and `envelope_state_count` will always be positive (i.e. some records with some size exist in the state), similar to the diff being always +1 because they are just inserts. 

But when upserting later on, the `stats.size_diff` or the `stats.values_diff` would represent diffs to the original value and in turn could be negative because maybe records were removed or were updated with smaller records taking up less space. We apply these diffs instead of setting the value directly.

File: src/storage/src/render/upsert/types.rs (hunk 12 above)
Comment (guswynn) [reply src/storage/src/render/upsert/types.rs:123]: This was you petros!! I just copied!

File: src/storage/src/render/upsert/types.rs (hunk 11 above)
Comment (guswynn) [reply src/storage/src/render/upsert/types.rs:93]: `PutValue` has 4 states, depending on if we are doing a delete or put, and if there was a previous value, or not!

`GetValue` certainly can be simplified!

File: src/rocksdb/src/lib.rs [hunk 14]
```diff
@@ -467,16 +514,21 @@ fn rocksdb_core_loop<K, V, M>(
...
                     match value {
                         Some(update) => {
                             encode_buf.clear();
-                            match bincode::serialize_into::<&mut Vec<u8>, _>(
-                                &mut encode_buf,
-                                &update,
-                            ) {
-                                Ok(()) => {}
+                            match enc_opts
+                                .serialize_into::<&mut Vec<u8>, _>(&mut encode_buf, &update)
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/rocksdb/src/lib.rs:529]: I don't want it to accidentally be serializing it into a `&mut [u8]` and have size issues, so I will leave it for clarity!

File: src/storage/src/render/upsert.rs (hunk 5 above)
Comment (guswynn) [reply src/storage/src/render/upsert.rs:316]: They happen only if their are bugs. Bugs in counting should not bring the impl down, which is why we warn (which goes to sentry) and default to 0 here

File: src/storage/src/render/upsert.rs (hunk 8 above)
Comment (guswynn) [reply src/storage/src/render/upsert.rs]: Moved into the consolidated impl in the new commit!

File: src/storage/src/render/upsert/rocksdb.rs (hunk 9 above)
Comment (guswynn) [reply src/storage/src/render/upsert/rocksdb.rs]: This is fixed in the newest commit!

File: src/storage/src/render/upsert/types.rs (hunk 13 above)
Comment (guswynn) [reply src/storage/src/render/upsert/types.rs:206]: Technically we will delete all the `0`'s as part of `merge_snapshot_chunk`, but this method doesnt know that, so ill fix the comment

File: src/storage/src/render/upsert/types.rs (hunk 1 above)
Comment (guswynn) [reply src/storage/src/render/upsert/types.rs:154]: updated!

File: src/rocksdb/src/lib.rs [hunk 15]
```diff
@@ -189,18 +189,46 @@ impl Options {
...
+/// The result type for `multi_get`.
+#[derive(Debug)]
+pub struct MultiGetResult {
+    /// The number of keys we fetched.
+    pub processed_gets: u64,
+}
+
+/// The result type for individual gets.
+#[derive(Debug, Default, Clone)]
+pub struct GetResult<V> {
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/rocksdb/src/lib.rs:197]: For the rocksdb crate I agree, but in `upsert`, we need to keep the value we are going to `put` back into the state, and the original size of the `get` (if it existed). Changing to your suggestion makes the `multi_get` impl for `rocksdb` a bit gross, but I updated the pr to do the change, so you can see!

**Comment by guswynn** [comment]
@petrosagg I update This pr to reflect all your comments, except for a few! A few notes:

> The XOR logic is leaking into each individual key/value implementation. That seems problematic to me as swapping the key/value store is no longer trivial or at least only concerned with the details of the store. Why does it have to be that way?

It was a perf thing for the hashmap impl, but I changed to this suggestion and its way better you are right

> I'm not sure if this [using `US::SNAPSHOT_SIZE` is worth doing.

I really want to have careful control over the batch sizes we sent to rocksdb (and eventually make it configurable), as it appears to be an important performance parameter, so I want to leave it as is

> Metric related things seem to have infiltrated every single type, carrying around decoded bytes, counts, and interleaving with the actual logic. This makes the code hard to follow, is there an alternative solution where we pass a Metrics registry or something to the implementations where they can set their metrics directly?

Hmm, there is some inherent complexity in the way we need `multi_get` and `multi_put` to manage sizes, but I may be able to move this into the `UpsertState` struct. I'll look at this today.

**Comment by guswynn** [comment]
@petrosagg Latest commit has the metrics moved down a level!

File: src/storage/src/render/upsert.rs [hunk 16]
```diff
@@ -320,18 +252,39 @@ where
     builder.build(move |caps| async move {
         let mut output_cap = caps.into_element();
 
-        let mut snapshot = vec![];
+        let mut state = UpsertState::new(
+            state().await,
+            upsert_shared_metrics,
+            upsert_metrics,
+            source_config.source_statistics,
```
Comment (moulimukherjee) [inline src/storage/src/render/upsert.rs:259]: dumb q: Since this is inside the operator, should this have been a `source_config.source_statistics.clone()` instead (which returns a `Rc`)? Why or why not?

Comment (guswynn) [reply src/storage/src/render/upsert.rs:259]: We already cloned it when we we made the `RawSourceConfig` we passed in. Here we are destructing the `source_config` and pulling out the `source_statistics` by value. In rust, you can do this if you only pull each field out at most once, and the struct doesn't implement `Drop`! It's subtle, not a dumb q!!

File: src/storage/src/render/upsert.rs (hunk 4 above)
Comment (petrosagg) [reply src/storage/src/render/upsert.rs:277]: I just tested this out locally and using `.now_or_never()` is pretty natural. I think this section should be with the XOR logic in here instead of wrapping the `UpsertStateBackend` in `UpsertState` because this logic is completely independent of what we use to store the data in. 

```rust
        let mut events = vec![];
        let mut snapshot_upper = Antichain::from_elem(Timestamp::minimum());
        while !PartialOrder::less_equal(&resume_upper, &snapshot_upper) {
            match previous.next_mut() {
                Some(AsyncEvent::Data(_cap, data)) => events.extend(data.drain(..)),
                Some(AsyncEvent::Progress(upper)) => snapshot_upper = upper,
                None => snapshot_upper = Antichain::new(),
            };
            while let Some(event) = previous.next_mut().now_or_never() {
                match event {
                    Some(AsyncEvent::Data(_cap, data)) => events.extend(data.drain(..)),
                    Some(AsyncEvent::Progress(upper)) => snapshot_upper = upper,
                    None => snapshot_upper = Antichain::new(),
                }
            }
            // 1. consolidate events in memory
            // 2. multi_get the values of all keys in `events`
            // 3. perform in-memory XOR
            // 4. multi_put the final in-memory state
        }
```

File: src/storage/src/render/upsert/types.rs [hunk 17]
```diff
@@ -54,85 +386,308 @@ impl Default for InMemoryHashMap {
...
     {
         let now = Instant::now();
-        let size = self.inner.multi_get(gets, results_out).await?;
+        let stats = self.inner.multi_get(gets, results_out).await?;
 
         self.metrics
             .multi_get_latency
             .observe(now.elapsed().as_secs_f64());
-        self.metrics.multi_get_size.observe(f64::cast_lossy(size));
+        self.metrics
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/render/upsert/types.rs:687]: The rocksdb crate has its own metrics for each chunk, these are across the entire unchunked `multi_get`

File: src/rocksdb/src/lib.rs [hunk 18]
```diff
@@ -364,9 +394,9 @@ where
 
         // We also unwrap all rocksdb errors here.
         match rx.await.map_err(|_| Error::RocksDBThreadGoneAway)? {
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/rocksdb/src/lib.rs:396]: It could be made a panic but these seemed cleaner to me!

File: src/rocksdb/src/lib.rs [hunk 19]
```diff
@@ -416,6 +448,8 @@ fn rocksdb_core_loop<K, V, M>(
     };
     let wo = options.as_rocksdb_write_options();
 
+    use bincode::Options;
+    let enc_opts = bincode::DefaultOptions::new().allow_trailing_bytes();
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/rocksdb/src/lib.rs:452]: ah yes, i missed this!

File: test/upsert/mzcompose.py [hunk 20]
```diff
@@ -102,3 +126,55 @@ def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
...
+            c.run("testdrive", "rehydration/02-source-setup.td")
+
+            c.kill("materialized")
+            c.kill("clusterd1")
+            c.up("materialized")
+            c.up("clusterd1")
+
+            c.run("testdrive", "rehydration/03-after-rehydration.td")
+
+        c.run("testdrive", "rehydration/04-reset.td")
```
Comment (moulimukherjee) [inline test/upsert/mzcompose.py:179]: Is the reset needed? It's empty

File: src/rocksdb/src/lib.rs [hunk 21]
```diff
@@ -441,21 +475,32 @@ fn rocksdb_core_loop<K, V, M>(
...
-                                Ok(v) => v,
-                                Err(e) => {
-                                    let _ = response_sender.send(Err(Error::DecodeError(e)));
-                                    return;
+                            let get_result = match previous_value {
+                                Some(previous_value) => {
+                                    match enc_opts.deserialize(&previous_value) {
+                                        Ok(value) => Some(GetResult {
+                                            value,
+
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/rocksdb/src/lib.rs:488]: how does rustfmt not get this?

File: test/upsert/mzcompose.py (hunk 20 above)
Comment (guswynn) [reply test/upsert/mzcompose.py:179]: Its the easiest way to get `testdrive` to run all the reset logic!

Comment (moulimukherjee) [reply test/upsert/mzcompose.py:179]: TIL! (I will try in my branch too, I was having trouble with resetting things)

**Review by petrosagg — APPROVED** [review]
talked about it with Gus offline (i.e online). there are a few things to change before merging but overall looks good!

File: src/timely-util/src/builder_async.rs [hunk 22]
```diff
@@ -187,6 +187,7 @@ impl<'handle, T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> Future
             return Poll::Ready(Some(Event::Progress(frontier.clone())));
         } else if frontier.is_empty() {
             // If the frontier is empty and is not pending it means that there is no more data left
+            println!("EMPTY FRONTIER");
```
Comment (danhhz) [inline src/timely-util/src/builder_async.rs:190]: @guswynn intentional?
